src/main/java/htsjdk/samtools/SAMSequenceDictionary.java

```diff
@@ -266,5 +268,100 @@ public String toString() {
+                throw new IllegalArgumentException(String.format("Non-equal sequence names (%s and %s) at index %d in " +
+                        "the dictionaries.",
+                        sequence.getSequenceName(), dict2.getSequence(sequenceIndex).getSequenceName(), sequenceIndex));
+            }
+
+            final Set<String> allTags = new HashSet<>(dict1.getSequence(sequenceIndex).getAttributes()
+                    .stream().map(Map.Entry::getKey).collect(Collectors.toSet()));
+            allTags.addAll(dict2.getSequence(sequenceIndex).getAttributes()
+                    .stream().map(Map.Entry::getKey).collect(Collectors.toSet()));
+
+            for (final String tag : allTags) {
+                final String value1 = dict1.getSequence(sequenceIndex).getAttribute(tag);
+                final String value2 = dict2.getSequence(sequenceIndex).getAttribute(tag);
+
+                if (value1 != null && value2 != null && !value1.equals(value2)) {
+                    if (tagsToMatch.contains(tag)) {
+                        final String error = String.format("Cannot merge the two dictionaries. Found sequence entry for which " +
+                                        "tag values differ: Sequence %s at tag %s has the values: %s and %s",
+                                dict1.getSequence(sequenceIndex).getSequenceName(),
+                                tag,
+                                value1,
+                                value2);
+
+                        log.error(error);
+                        throw new IllegalArgumentException(error);
+                    } else {
+                        log.warn("Found sequence entry for which " +
+                                        "tags differ: %s and tag %s has the two values: %s and %s. Using Value %s",
+                                dict1.getSequence(sequenceIndex).getSequenceName(),
+                                tag,
+                                value1,
+                                value2, value1);
+                    }
+                }
+                mergedSequence.setAttribute(tag, value1 == null ? value2 : value1);
+            }
+
+            final int length1 = dict1.getSequence(sequenceIndex).getSequenceLength();
+            final int length2 = dict2.getSequence(sequenceIndex).getSequenceLength();
+
+            if (length1 != UNKNOWN_SEQUENCE_LENGTH && length2 != UNKNOWN_SEQUENCE_LENGTH && length1 != length2) {
+                if (tagsToMatch.contains(SAMSequenceRecord.SEQUENCE_LENGTH_TAG)) {
+                    final String error = String.format("Cannot merge the two dictionaries. Found sequence entry for which " +
+                                    "lengths differ: %s has lengths %s and %s",
+                            dict1.getSequence(sequenceIndex).getSequenceName(),
+                            length1, length2);
+
+                    log.error(error);
+                    throw new IllegalArgumentException(error);
```

Comment:
Logging and then throwing is just going to lead to the same error getting printed multiple times.  Please pick one!